### ISSUE TITLE
Assigning user assigned identity to Pool

### ISSUE THREAD

**Issue by lolochristen** [Azure/Batch#114]
#### Problem Description
To be able to user assigned identities to consume resource on Azure, I want to assign such an identity to a pool using ClousPool.Identity property. The Identity property is null and cannot be assigned because BatchPoolIdentity has only internal constructors. Is there another way to achieve this?

#### Steps to Reproduce
```cs
pool = batchClient.PoolOperations.CreatePool(...);
pool.Identity = new BatchPoolIdentity(PoolIdentityType.UserAssigned);
pool.Identity.UserAssignedIdentities = new List<UserAssignedIdentity>()  {   new(....)  };
```

#### Expected Results
CloudPool as a assigned managed identity and uses this to consume Azure Resource/Storage Accounts.

#### Actual Results
CloudPool.Identity cannot be set as BatchPoolIdentity is internal

**Comment by staer** [comment]
Are you using BatchClient or BatchManagementClient? When assigning identities I believe you need to use the BatchManagementClient, the regular BatchClient will only fill in the identity information on a GET which is why is is unsettable. Identities must be set through the ARM API which is what the BatchManagementClient is using under the hood.

**Comment by lolochristen** [comment]
I am using BatchClient and no BatchManagementClient exists in the SDK and don't see the need to prevent setting Identities as the SDK supports to create pools and under the hood it just consumes the Batch Rest Api which supports setting identities. I require to create the pools dynamically on demand.

**Comment by staer** [comment]
Sorry for the delayed response, I was out of the office. Give the management SDK a try: 

https://www.nuget.org/packages/Microsoft.Azure.Management.Batch/

Under the hood, there are two SDK's available for Batch. The "normal" SDK uses the batch REST API, while the management SDK uses the more generalized ARM sdk. Identities can only be set using the ARM SDK due to underlying Azure Platform features. Using the management SDK will still allow you to create pools dynamically on-demand, it's just hitting a different endpoint within Azure.

**Comment by martinjoshua** [comment]
Will the ability to specify a user managed identity be added to the Batch SDK?

The Batch Management SDK is not an option for our scenario because according to the documentation,[ administrator or coadministrator access to the subscription is required](https://docs.microsoft.com/en-us/azure/batch/batch-management-dotnet). The application that we are building must only have access to create pools (and specify their identity), not subscription-level access.

**Comment by dpwatrous** [comment]
@martinjoshua As far as I know we are unable to support managed identity assignment through our data plane SDK because managed identities are a feature of Azure Resource Manager (ARM), so API calls must go through an ARM endpoint rather than directly to the Batch service endpoints.

I suspect that documentation is out of date, as "Administrator" and "Co-Administrator" are roles that predate the current Azure Active Directory RBAC system. The SDK will need the permissions required to create a Batch pool through Azure Resource Manager, and also the 'Managed Identity Operator' permission to assign a managed identity, but it shouldn't require any kind of global admin permissions to your subscription. [See the list of built-in RBAC roles for more info](https://docs.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#identity).

I'll follow up with the team to confirm the docs are incorrect here and make sure we update them. We may just end up removing that section as there isn't any single specific role that the management SDK requires, but rather varying sets of permissions depending on what operations you perform with the SDK.

**Comment by martinjoshua** [comment]
Thank you @dpwatrous for the clarification! You are correct. Once I went refactored to our code to use the Management SDK and set the appropriate permissions as you suggested, everything worked. Thanks!!

Unfortunately, switching over to the managed identity introduced a new problem: after running for ~10 minutes or so, we started receiving a 429 error when obtaining a managed identity token. This is likely caused by all the `ResourceFile.FromStorageContainerUrl(...)` and `new OutputFileBlobContainerDestination(...)` where we now include a `ComputeNodeIdentityReference` configured to our managed identity. Does batch do any token caching when ResourceFile and OutputFileBlobContainerDestination objects are instantiated in this manner?

**Comment by martinjoshua** [comment]
@dpwatrous one idea that I had for a potential work-around the 429 issue is to use the JobStartTask to run a series of shell commands to obtain a token, then set an environment variable to a SAS generated from that token. With that, one could embed the environment variable into the containerURL when using `ResourceFile.FromStorageContainerUrl(...)` and `new OutputFileBlobContainerDestination(...)`

Unfortunately that did not work, as environment variables do not get expanded in the containerURL (like the path parameter https://github.com/Azure/Batch/issues/130)

Is there a way to force each node to cache managed identity tokens?

**Comment by staer** [comment]
@martinjoshua How many files are you downloading? This is an issue that has already been fixed in the batch agent and is pending release. As a workaround for now I would suggest using:

https://docs.microsoft.com/en-us/azure/batch/resource-files#storage-container-url

And downloading files in bulk from a container with a prefix instead of using many resource files. Alternatively, you can fall back on SAS until the release happens. I will update thread when release commences.

**Comment by martinjoshua** [comment]
Thanks @staer, SAS will work for now. Glad to hear that this has already been addressed!

The number of files that are downloaded varies between 10 - 30, however, there are never more than 5 ResourceFile objects to a single CloudTask; we use a prefix as much as possible. I wonder if we hit the threshold from the output files? Some tasks have a `OutputFileBlobContainerDestination(...)` that use a filePattern, which will sometimes match around 180 files. If these tasks were running concurrently, I can imagine us hitting the IMDS threshold. When the filePattern is used, does Batch use the same token for each file or obtain a new one every time?

**Comment by staer** [comment]
Without disclosing too much about the nature of the bug itself, the batch agent uses a cache for all identity tokens however there was a case where if certain code paths happened at the same time the cache was invalidated and the IMDS endpoint would be hit in error which could cause 429's in some cases.

**Comment by staer** [comment]
and yes, i suspect (without going into the code) that a filepattern would cause this bad interaction

**Comment by martinjoshua** [comment]
Thank you for confirming, I will just hold tight until the next batch agent release!

**Comment by martinjoshua** [comment]
Hello @staer, I noticed you said in another thread that a new version of the batch agent has been released. Does it include the bug discussed above? Thanks!

**Comment by staer** [comment]
@martinjoshua Yes! I knew there is another issue that referenced this, i just couldn't find it because it was already closed :)

**Comment by martinjoshua** [comment]
Ok awesome - thank you!
